File: app/src/main/java/pl/pola_app/model/PhotoPicturesSendData.java
package pl.pola_app.model;



import java.util.LinkedList;
import java.util.List;

import rx.Observable;

public class PhotoPicturesSendData{

    private final int width;
    private final int height;

    private final List<FileToSendData> fileToSendDataList = new LinkedList<>();

    PhotoPicturesSendData(int width, int height) {
        this.width = width;
        this.height = height;
    }

    public Observable<FileToSendData> createFileToSendDataObservable() {
        return Observable.from(fileToSendDataList);
    }

    public static PhotoPicturesSendData create(PhotoPicturesResult photoPicturesResult, List<TakenPicInfo> files) {
        final int size = photoPicturesResult.signedRequestsSize();
        if(size != files.size()) {
            throw new IllegalArgumentException("PhotoPicturesResult is not compatible with file list - different size!");
        }
        final PhotoPicturesSendData result = new PhotoPicturesSendData(photoPicturesResult.width, photoPicturesResult.height);
        for(int index = 0; index < size; index ++) {
            result.fileToSendDataList.add(new FileToSendData(files.get(index), photoPicturesResult.getSignedRequestAt(index)));
        }
        return result;

    }
}

File: app/src/main/java/pl/pola_app/model/PhotoPicturesSendData.kt
package pl.pola_app.model

import rx.Observable
import java.util.*

data class PhotoPicturesSendData internal constructor(private val width: Int, private val height: Int) {
    private val fileToSendDataList: MutableList<FileToSendData> = LinkedList()
    fun createFileToSendDataObservable(): Observable<FileToSendData> {
        return Observable.from(fileToSendDataList)
    }

    companion object {
        fun create(
            photoPicturesResult: PhotoPicturesResult,
            files: List<TakenPicInfo>
        ): PhotoPicturesSendData {
            val size = photoPicturesResult.signedRequestsSize()
            require(size == files.size) { "PhotoPicturesResult is not compatible with file list - different size!" }
            val result =
                PhotoPicturesSendData(photoPicturesResult.width, photoPicturesResult.height)
            for (index in 0 until size) {
                result.fileToSendDataList.add(
                    FileToSendData(
                        files[index],
                        photoPicturesResult.getSignedRequestAt(index)
                    )
                )
            }
            return result
        }
    }
}